File: common/src/main/java/generations/gg/generations/core/generationscore/api/player/PlayerMoneyHandler.java
package generations.gg.generations.core.generationscore.api.player;

import net.minecraft.world.entity.player.Player;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class PlayerMoneyHandler {
    public static final Map<Player, PlayerMoney> MONEY_CACHE = new HashMap<>();


    private static Function<Player, PlayerMoney> function = PlayerPokeDollars::new;

    public static void setFunction(Function<Player, PlayerMoney> function) {
        PlayerMoneyHandler.function = function;
    }

    public static PlayerMoney of(Player player) {
        return MONEY_CACHE.computeIfAbsent(player, function);
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/api/player/PlayerMoneyHandler.kt
package generations.gg.generations.core.generationscore.common.api.player

import net.minecraft.world.entity.player.Player
import java.util.function.Function

object PlayerMoneyHandler {
    val MONEY_CACHE: MutableMap<Player, PlayerMoney> = HashMap()


    private var function =
        Function<Player, PlayerMoney> { player: Player ->
            PlayerPokeDollars(
                player
            )
        }

    fun setFunction(function: Function<Player, PlayerMoney>) {
        PlayerMoneyHandler.function = function
    }

    @JvmStatic
    fun of(player: Player): PlayerMoney {
        return MONEY_CACHE.computeIfAbsent(player, function)
    }
}
